parsons/salesforce/salesforce.py

```diff
@@ -12,33 +13,71 @@ class Salesforce:
             Settings > My Personal Information > Reset My Security Token.
-            Not required if ``SALESFORCE_SECURITY_TOKEN`` env variable is passed.
+            Not required if ``SALESFORCE_SECURITY_TOKEN`` env variable is passed. Used in the 'password' auth method.
         test_environment: bool
             If ``True`` the client will connect to a Salesforce sandbox instance. Not required if
             ``SALESFORCE_DOMAIN`` env variable is passed.
+        consumer_key: str
+            consumer key for a connected app. Used in the 'client_credentials' auth method.
+        consumer_secret: str
+            consumer secret for a connected app. Used in the 'client_credentials' auth method.
+        domain: str
+            url for the salesforce instance. Used in the 'client_credentials' auth method
+        authentication_method: str
+            the method to use for authentication. defaults to "password". Not required if ``SALESFORCE_AUTHENTICATION_METHOD`` env variable is passed.
+
     `Returns:`
         Salesforce class
     """
 
-    def __init__(self, username=None, password=None, security_token=None, test_environment=False):
-        self.username = check_env.check("SALESFORCE_USERNAME", username)
-        self.password = check_env.check("SALESFORCE_PASSWORD", password)
-        self.security_token = check_env.check("SALESFORCE_SECURITY_TOKEN", security_token)
+    def __init__(
+        self,
+        username=None,
+        password=None,
+        security_token=None,
+        test_environment=False,
+        consumer_key=None,
+        consumer_secret=None,
+        domain=None,
+        authentication_method=None,
+    ):
+        if authentication_method:
+            self.authentication_method = authentication_method
+        elif env_authentication_method := os.environ.get("SALESFORCE_AUTHENTICATION_METHOD"):
+            self.authentication_method = env_authentication_method
+        else:
+            self.authentication_method = "password"
+
+        if self.authentication_method == "password":
+            self.username = check_env.check("SALESFORCE_USERNAME", username)
+            self.password = check_env.check("SALESFORCE_PASSWORD", password)
+            self.security_token = check_env.check("SALESFORCE_SECURITY_TOKEN", security_token)
+            if test_environment:
+                self.domain = check_env.check("SALESFORCE_DOMAIN", "test")
+            else:
+                self.domain = None
+
+        elif self.authentication_method == "client_credentials":
+            self.consumer_key = check_env.check("SALESFORCE_CONSUMER_KEY", consumer_key)
+            self.consumer_secret = check_env.check("SALESFORCE_CONSUMER_SECRET", consumer_key)
+            self.domain = check_env.check("SALESFORCE_DOMAIN", domain)
 
-        if test_environment:
-            self.domain = check_env.check("SALESFORCE_DOMAIN", "test")
         else:
-            self.domain = None
+            raise NotImplementedError(
```

Comment:
this will be raise when self.authentication_method is a string, but not `client_credentials` or `password`